mm2src/coins/nft/trading_proto_v2/mod.rs

```diff
@@ -0,0 +1,549 @@
+        }
+    }
+
+    pub(crate) async fn refund_nft_maker_payment_v2_timelock_impl(
+        &self,
+        args: RefundPaymentArgs<'_>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let _etomic_swap_contract = try_tx_s!(args.swap_contract_address.try_to_address());
+        let tx: UnverifiedTransaction = try_tx_s!(rlp::decode(args.payment_tx));
+        let _payment = try_tx_s!(SignedEthTx::new(tx));
+        todo!()
+    }
+
+    fn prepare_nft_maker_payment_v2_data(
+        &self,
+        contract_type: &ContractType,
+        swap_contract_address: Address,
+        token_id: U256,
+        args: &SendNftMakerPaymentArgs<'_, Self>,
+        htlc_data: Vec<u8>,
+    ) -> Result<Vec<u8>, PrepareTxDataError> {
+        match contract_type {
+            ContractType::Erc1155 => {
+                let function = ERC1155_CONTRACT.function("safeTransferFrom")?;
+                let amount_u256 = U256::from_dec_str(&args.amount.to_string())
+                    .map_err(|e| PrepareTxDataError::Internal(e.to_string()))?;
+                let data = function.encode_input(&[
+                    Token::Address(self.my_address),
+                    Token::Address(swap_contract_address),
+                    Token::Uint(token_id),
+                    Token::Uint(amount_u256),
+                    Token::Bytes(htlc_data),
+                ])?;
+                Ok(data)
+            },
+            ContractType::Erc721 => {
+                let function = self.erc721_transfer_with_data()?;
+                let data = function.encode_input(&[
+                    Token::Address(self.my_address),
+                    Token::Address(swap_contract_address),
+                    Token::Uint(token_id),
+                    Token::Bytes(htlc_data),
+                ])?;
+                Ok(data)
+            },
+        }
+    }
+
+    fn validate_payment_args<'a>(
+        &self,
+        taker_secret_hash: &'a [u8],
+        maker_secret_hash: &'a [u8],
+        amount: &BigDecimal,
+        contract_type: &ContractType,
+    ) -> Result<(), String> {
+        match contract_type {
+            ContractType::Erc1155 => {
+                if !is_positive_integer(amount) {
+                    return Err("ERC-1155 amount must be a positive integer".to_string());
+                }
+            },
+            ContractType::Erc721 => {
+                if amount != &BigDecimal::from(1) {
+                    return Err("ERC-721 amount must be 1".to_string());
+                }
+            },
+        }
+        if taker_secret_hash.len() != 32 {
+            return Err("taker_secret_hash must be 32 bytes".to_string());
+        }
+        if maker_secret_hash.len() != 32 {
+            return Err("maker_secret_hash must be 32 bytes".to_string());
+        }
+
+        Ok(())
+    }
+
+    fn prepare_htlc_data(
+        &self,
+        args: &SendNftMakerPaymentArgs<'_, Self>,
+        taker_address: Address,
+        token_address: Address,
```

Comment:
`token_address` can be retrieved from args, no need for an additional parameter.

Reply:
fixed here https://github.com/KomodoPlatform/komodo-defi-framework/commit/8b5476b8d55aeebfb5cf38aebf8757c9f75b0090